jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3614,6 +3656,41 @@ private static final void validateArgumentsOfCountedLoop(MethodHandle startHandl
 				/*[MSG "K065I", "The 2nd parameter type of the body handle with a non-void return type must be int rather than {0}"]*/
 				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065I", bodyType.toMethodDescriptorString())); //$NON-NLS-1$
 			}
+			
+			bodyIterationVarLength = 2;
+		}
+
+		/* As a special case, if the body contributes only V and I types, with no
+		 * additional A types, then the internal parameter list is extended by the
+		 * argument types A... of the end handle. Adjust A count to correctly verify
+		 * parameters are effectively identical. Body will be altered outside of
+		 * verification.
+		 */
+		int externalParamLength = bodyParamLength - bodyIterationVarLength;
+		if (bodyParamLength == bodyIterationVarLength) {
+			externalParamLength = endHandle.type.parameterCount();
+		}
+		
+		/* Verify that handles are effectively identical and constrained by body. All we
+		 * need to do here is verify that no parameters have longer argument lists than
+		 * body. The generic loop will compare arguments to ensure all clauses are
+		 * effectively identical.
+		 */
+		if (startHandle.type.parameterCount() > externalParamLength) {
+			/*[MSG "K0654", "The count of {0}'s parameters must be less than the external parameters."]*/
+			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0654", "start handle"));  //$NON-NLS-2$
+		}
+		
+		if (endHandle.type.parameterCount() > externalParamLength) {
+			/*[MSG "K0654", "The count of {0}'s parameters must be less than the external parameters."]*/
+			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0654", "end handle"));  //$NON-NLS-2$
+		}
+		
+		if (null != initHandle) {
+			if (initHandle.type.parameterCount() > externalParamLength) {
```

Comment:
Better merge to one statement like:
if ((null != initHandle) && (initHandle.type.parameterCount() > externalParamLength)) {
...
}